src/drivers/net/usbnet.c

```diff
@@ -0,0 +1,212 @@
+
+	in_endp = nic_priv->usbdev->endpoints[2];
+	assert(in_endp);
+
+	switch(req->req_stat) {
+	case USB_REQ_NOERR:
+		memcpy(nic_priv->pdata, req->buf, req->len);
+		nic_priv->pdata += req->len;
+		break;
+	default:
+		return;
+	}
+
+	/* End of packet */
+	if (req->len < in_endp->max_packet_size) {
+		struct sk_buff *skb;
+		size_t len;
+
+		len = nic_priv->pdata - nic_priv->data;
+		if ((skb = skb_alloc(len))) {
+			memcpy(skb->mac.raw, nic_priv->data, len);
+			skb->dev = nic;
+			netif_rx(skb);
+		} else {
+			printk("usbnet: skbuff allocation failed\n");
+		}
+
+		/* Prepare for copying new packet */
+		nic_priv->pdata = nic_priv->data;
+
+		timer_init_msec(&nic_priv->timer, TIMER_ONESHOT,
+				USBNET_TIMER_FREQ, usbnet_timer_handler, nic_priv);
+	} else {
+		/* Receive the next part of the packet */
+		usb_endp_bulk(in_endp, usbnet_rcv_notify, nic_priv->pdata,
+				in_endp->max_packet_size);
+	}
+}
+
+static void usbnet_timer_handler(struct sys_timer *tmr, void *param) {
+	struct usb_endp *in_endp;
+	struct usbnet_priv *nic_priv = (struct usbnet_priv *) param;
+
+	in_endp = nic_priv->usbdev->endpoints[2];
+	assert(in_endp);
+
+	usb_endp_bulk(in_endp, usbnet_rcv_notify, nic_priv->pdata,
+			in_endp->max_packet_size);
+}
+
+static void usbnet_iface_hnd(struct usb_request *req, void *arg) {
+}
+
+static int usbnet_probe(struct usb_driver *drv, struct usb_dev *dev,
+		void **data) {
+	struct net_device *nic;
+	struct usbnet_priv *nic_priv;
+	struct usb_class_cdc *cdc = usb2cdcdata(dev);
+	int res;
+
+	/* Enable in/out DATA interface */
+	cdc_set_interface(dev, 1, 1, usbnet_iface_hnd);
+
+	nic = (struct net_device *) etherdev_alloc(sizeof *nic_priv);
+	if (!nic) {
+		return -ENOMEM;
+	}
+	nic->drv_ops = &usbnet_drv_ops;
+	nic_priv = netdev_priv(nic, struct usbnet_priv);
+	nic_priv->usbdev = dev;
+	nic_priv->data = nic_priv->pdata = sysmalloc(ETH_FRAME_LEN);
+	if (!nic_priv->data) {
+		etherdev_free(nic);
+		return -ENOMEM;
+	}
+
+	cdc->privdata = (void *) nic;
+
+	res = inetdev_register_dev(nic);
+	if (res < 0) {
+		goto err_out;
+	}
+
+    timer_init_msec(&nic_priv->timer, TIMER_ONESHOT,
+            USBNET_TIMER_FREQ, usbnet_timer_handler, nic_priv);
+	if (res < 0) {
+		goto err_out;
+	}
+
+    return 0;
+
+err_out:
+	sysfree(nic_priv->data);
+	etherdev_free(nic);
+	return res;
+}
+
+static void usbnet_disconnect(struct usb_dev *dev, void *data) {
+}
+
+static struct usb_device_id usbnet_id_table[] = {
+	{ 0x0525, 0xa4a2 },
+	{ },
+};
+
+static void usbnet_send_notify_hnd(struct usb_request *req, void *arg) {
+}
+
+static void usb_net_bulk_send(struct usb_dev *dev, struct sk_buff *skb) {
+	struct usb_endp *endp;
+	size_t i = 0;
+	int len;
+
+	endp = dev->endpoints[3];
+
+	for (i = skb->len, len = min(skb->len, endp->max_packet_size);
+			i != 0; i -= len, len = min(i, endp->max_packet_size)) {
+		usb_endp_bulk(endp, usbnet_send_notify_hnd,
+				skb->mac.raw + (skb->len - i), len);
+	}
+
+	if (len == endp->max_packet_size) {
```

Comment:
Ah, I see now...

Then, I guess, this will work again as is after changing the above loop to
```c
for (i = skb->len; i != 0 && (len = min(i, endp->max_packet_size)); i -= len)
    ...
```
or to a `while` version, which I've suggested earlier.